services/job_metrics.go

```diff
@@ -0,0 +1,129 @@
+package services
+
+import (
+	"expvar"
+	"github.com/smartcontractkit/chainlink/store"
+	"github.com/smartcontractkit/chainlink/store/models"
+	"go.uber.org/multierr"
+)
+
+// JobMetrics safely stores all job metric data in memory
+type JobMetrics interface {
+	Start() error
+	Add(models.JobSpec) error
+	AddRun(models.JobRun)
+	Get(string) models.JobSpecMetrics
+}
+
+type jobMetrics struct {
+	collection map[string]models.JobSpecMetrics
+	eMap       *expvar.Map
+	store      *store.Store
+}
+
+// NewJobMetrics initializes a JobMetrics.
+func NewJobMetrics(str *store.Store) JobMetrics {
+	n := "jobmetrics"
+	if expvar.Get(n) == nil {
+		expvar.NewMap(n)
+	}
+	return &jobMetrics{
+		collection: make(map[string]models.JobSpecMetrics),
+		eMap:       expvar.Get(n).(*expvar.Map),
+		store:      str,
+	}
+}
+
+// Start fetches all the current jobs and their runs from the store
+// then adds them to the collection
+func (jm *jobMetrics) Start() error {
+	var merr error
+
+	aj, err := jm.store.Jobs()
+	if err != nil {
```

Comment:
This might bog down the system, having to iterate over every single job run on boot. Storing every job in memory with metrics could be ok, at least we're not storing every job run in memory.

Reply:
A way around this would be to write and store these aggregate metrics as part of our normal operation `SaveRun`/`SaveJob` methods. Caching on write rather than calculating on boot.

Reply:
Right 👍 

We also need to figure out the viability of holding every job in memory. Can a production chainlink node hold N jobs in memory? Maybe. JobRuns will grow far faster. Just something to think about it.

Reply:
@jleeh I think this raises an interesting question: do you actually need all jobs?

Could we return the last 100 and allow you to paginate if you need more?

This way, we could write the metrics as they occur, and then simple get them from bolt when a GET reaches this endpoint. This wouldn’t be supported out the box by ‘expvar’ which is a bummer. I’m enjoying expvar but it seems better suited for global counts of constant size rather than an O(N) map of jobs.